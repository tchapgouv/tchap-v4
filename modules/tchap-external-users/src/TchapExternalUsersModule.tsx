/*
Copyright 2022 New Vector Ltd. t/a Element

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

import { RuntimeModule } from "@matrix-org/react-sdk-module-api/lib/RuntimeModule";
import { ModuleApi } from "@matrix-org/react-sdk-module-api/lib/ModuleApi";
import { CantAddDialog  } from "./components/CantAddDialog";
import { AddRoomListener, ExternalUsersLifecycle } from "@matrix-org/react-sdk-module-api/lib/tchap/lifecycles/ExternalUsers";
import { DialogProps } from "@matrix-org/react-sdk-module-api/lib/components/DialogContent";

import React from "react";

export default class TchapExternalUsersModule extends RuntimeModule {
    private domain: String;

    public constructor(moduleApi: ModuleApi) {
        super(moduleApi);

        // const rawConfig = this.moduleApi.getConfigValue(
        //     TCHAP_EXTERNAL_USERS_MODULE_CONFIG_NAMESPACE,
        //     TCHAP_EXTERNAL_USERS_MODULE_CONFIG_KEY,
        //   );
      
        //   assertValidGuestModuleConfig(rawConfig);
        //   this.config = rawConfig;

        this.on(ExternalUsersLifecycle.AddRoom, this.onAddRoom);
    }

    protected onAddRoom: AddRoomListener = async () => {
        await this.moduleApi.openDialog<null, DialogProps, CantAddDialog>("test", (props, ref) => <CantAddDialog ref={ref} {...props} />);
    };
}
