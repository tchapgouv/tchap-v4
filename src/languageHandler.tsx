/*
Copyright 2024 New Vector Ltd.
Copyright 2023 The Matrix.org Foundation C.I.C.

SPDX-License-Identifier: AGPL-3.0-only OR GPL-3.0-only
Please see LICENSE files in the repository root for full details.
*/

import React from "react";
import {
    IVariables,
    TranslatedString,
    TranslationKey as ReactTranslationKey,
    // eslint-disable-next-line camelcase
    _t as react_t,
    // eslint-disable-next-line camelcase
    _td as react_td,
    // eslint-disable-next-line camelcase
    _tDom as react_tDom,
    Tags,
    UserFriendlyError as ReactUserFriendlyError,
    ErrorOptions,
} from "matrix-react-sdk/src/languageHandler";
import { Leaves } from "matrix-react-sdk/src/@types/common";

import type ReactEN from "matrix-react-sdk/src/i18n/strings/en_EN.json";
import type EN from "./i18n/strings/en_EN.json";

/**
 * This module wraps languageHandler in the matrix-react-sdk and adds type casts to include translations
 * which we know will be injected by webpack.
 */

export type TranslationKey = Leaves<typeof EN & typeof ReactEN, "|", string | { other: string }, 4>;

export class UserFriendlyError extends ReactUserFriendlyError {
    public constructor(message: TranslationKey, substitutionVariablesAndCause?: IVariables & ErrorOptions) {
        super(message as ReactTranslationKey, substitutionVariablesAndCause);
    }
}

export function _td(s: TranslationKey): TranslationKey {
    return react_td(s as ReactTranslationKey);
}

// eslint-next-line @typescript-eslint/naming-convention
export function _t(text: TranslationKey, variables?: IVariables): string;
export function _t(text: TranslationKey, variables: IVariables | undefined, tags: Tags): React.ReactNode;
export function _t(text: TranslationKey, variables?: IVariables, tags?: Tags): TranslatedString {
    return react_t(text as ReactTranslationKey, variables, tags!);
}

// eslint-next-line @typescript-eslint/naming-convention
export function _tDom(text: TranslationKey, variables?: IVariables): TranslatedString;
export function _tDom(text: TranslationKey, variables: IVariables, tags: Tags): React.ReactNode;
export function _tDom(text: TranslationKey, variables?: IVariables, tags?: Tags): TranslatedString {
    return react_tDom(text as ReactTranslationKey, variables!, tags!);
}
