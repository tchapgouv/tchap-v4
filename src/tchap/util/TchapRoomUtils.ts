/**
 * Tchap Room utils.
 */

import { Room } from "matrix-js-sdk/src/matrix";
import { MatrixClientPeg } from "matrix-react-sdk/src/MatrixClientPeg";
import { ISendEventResponse } from "matrix-js-sdk/src/@types/requests";
import * as utils from "matrix-js-sdk/src/utils";
import { Method } from "matrix-js-sdk/src/http-api";

import { TchapRoomAccessRule, TchapRoomAccessRulesEventId, TchapRoomType } from "../@types/tchap";

export default class TchapRoomUtils {
    //inspired by https://github.com/tchapgouv/tchap-android/blob/develop/vector/src/main/java/fr/gouv/tchap/core/utils/RoomUtils.kt#L31
    //direct type is not handled yet
    static getTchapRoomType(room: Room): TchapRoomType {
        const isEncrypted: boolean = this.isRoomEncrypted(room.roomId);
        const tchapRoomAccessRule: TchapRoomAccessRule = this.getTchapRoomAccessRule(room);
        return this.getTchapRoomTypeInternal(isEncrypted, tchapRoomAccessRule);
    }

    static getTchapRoomTypeInternal(isEncrypted: boolean, tchapRoomAccessRule: TchapRoomAccessRule): TchapRoomType {
        if (!isEncrypted) {
            return TchapRoomType.Forum;
        }
        if (!tchapRoomAccessRule) {
            return TchapRoomType.Unknown;
        }
        if (tchapRoomAccessRule === TchapRoomAccessRule.Restricted) {
            return TchapRoomType.Private;
        }
        if (tchapRoomAccessRule === TchapRoomAccessRule.Unrestricted) {
            return TchapRoomType.External;
        }
        return TchapRoomType.Unknown;
    }

    /**
     * Get Tchap Room Access Rule
     * @param room
     * @returns string that matches of one TchapRoomAccessRule //todo or null? or empty?
     */
    static getTchapRoomAccessRule(room: Room): TchapRoomAccessRule {
        return room.currentState.getStateEvents(TchapRoomAccessRulesEventId, "")?.getContent().rule;
    }

    /**
     * Is room encrypted? search in local storage or in server if needed
     * @param roomId
     * @returns true if room is encrypted, false if not
     */
    static isRoomEncrypted(roomId: string): boolean {
        return MatrixClientPeg.get().isRoomEncrypted(roomId);
    }

    /**
     * Update public room join rule to "invite" state (so that the room be private).
     * @returns Promise which resolves: to an ISendEventResponse object
     * @returns Rejects: with an error response.
     */
    static updateRoomJoinRuleToInvite(
        roomId: string,
    ): Promise<ISendEventResponse> {
        const client = MatrixClientPeg.get();
        const content = { join_rule: 'invite' };
        const path = utils.encodeUri("/rooms/$roomId/state/m.room.join_rules", {
            $roomId: roomId,
        });

        return client.http.authedRequest(Method.Put, path, undefined, content);
    }

    /**
     * Turn on the encryption in a room (if this is not already done).
     * @returns Promise which resolves: to an ISendEventResponse object
     * @returns Rejects: with an error response.
     */
    static turnOnEncryptionInThisRoom(
        roomId: string,
    ): Promise<ISendEventResponse> {
        const client = MatrixClientPeg.get();
        const content = { algorithm: 'm.megolm.v1.aes-sha2' };
        const path = utils.encodeUri("/rooms/$roomId/state/m.room.encryption", {
            $roomId: roomId,
        });

        return client.http.authedRequest(Method.Put, path, undefined, content);
    }

    /**
     * Remove a room from the room directory.
     * @returns Promise which resolves: to an ISendEventResponse object
     * @returns Rejects: with an error response.
     */
    static removeRoomFromRoomDirectory(
        roomId: string,
    ): Promise<ISendEventResponse> {
        const client = MatrixClientPeg.get();
        const content = { visibility: 'private' };
        const path = utils.encodeUri("/rooms/$roomId/state/m.room.history_visibility", {
            $roomId: roomId,
        });

        return client.http.authedRequest(Method.Put, path, undefined, content);
    }
}
