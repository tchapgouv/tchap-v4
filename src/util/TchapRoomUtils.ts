/**
 * Tchap Room utils.
 */

import { Room } from "matrix-js-sdk/src/matrix";
import { MatrixClientPeg } from "matrix-react-sdk/src/MatrixClientPeg";

import { TchapRoomAccessRule, TchapRoomAccessRulesEventId, TchapRoomType } from "../tchap/@types/tchap";

export default class TchapRoomUtils {
    //inspired by https://github.com/tchapgouv/tchap-android/blob/develop/vector/src/main/java/fr/gouv/tchap/core/utils/RoomUtils.kt#L31
    //direct type is not handled yet
    static getTchapRoomType(room: Room): TchapRoomType {
        const isEncrypted: boolean = this.isRoomEncrypted(room.roomId);
        const tchapRoomAccessRule: TchapRoomAccessRule = this.getTchapRoomAccessRule(room);
        return this.getTchapRoomTypeInternal(isEncrypted, tchapRoomAccessRule);
    }

    static getTchapRoomTypeInternal(isEncrypted: boolean, tchapRoomAccessRule: TchapRoomAccessRule): TchapRoomType {
        if (!isEncrypted) {
            return TchapRoomType.Forum;
        }
        if (!tchapRoomAccessRule) {
            return TchapRoomType.Unknown;
        }
        if (tchapRoomAccessRule === TchapRoomAccessRule.Restricted) {
            return TchapRoomType.Private;
        }
        if (tchapRoomAccessRule === TchapRoomAccessRule.Unrestricted) {
            return TchapRoomType.External;
        }
        return TchapRoomType.Unknown;
    }

    /**
     * Get Tchap Room Access Rule
     * @param room
     * @returns string that matches of one TchapRoomAccessRule //todo or null? or empty?
     */
    static getTchapRoomAccessRule(room: Room): TchapRoomAccessRule {
        return room.currentState.getStateEvents(TchapRoomAccessRulesEventId, "")?.getContent().rule;
    }

    /**
     * Is room encrypted? search in local storage or in server if needed
     * @param roomId
     * @returns true if room is encrypted, false if not
     */
    static isRoomEncrypted(roomId: string): boolean {
        return MatrixClientPeg.get().isRoomEncrypted(roomId);
    }
}
