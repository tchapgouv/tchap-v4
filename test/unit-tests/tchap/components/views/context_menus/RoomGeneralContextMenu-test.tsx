import { logRoles, render, screen } from "jest-matrix-react";
import { mocked } from "jest-mock";
import { MatrixClient, PendingEventOrdering, Room, JoinRule, KnownMembership } from "matrix-js-sdk/src/matrix";
import React from "react";

import { stubClient } from "~tchap-web/test/test-utils/test-utils";
import { clearAllModals } from "~tchap-web/test/test-utils";
import { ChevronFace } from "~tchap-web/src/components/structures/ContextMenu";
import {
    RoomGeneralContextMenu,
    RoomGeneralContextMenuProps,
} from "~tchap-web/src/components/views/context_menus/RoomGeneralContextMenu";
import MatrixClientContext from "~tchap-web/src/contexts/MatrixClientContext";
import { MatrixClientPeg } from "~tchap-web/src/MatrixClientPeg";
import { DefaultTagID } from "~tchap-web/src/stores/room-list/models";
import RoomListStore from "~tchap-web/src/stores/room-list/RoomListStore";
import DMRoomMap from "~tchap-web/src/utils/DMRoomMap";
import TchapRoomUtils from "~tchap-web/src/tchap/util/TchapRoomUtils";

jest.mock("~tchap-web/src/tchap/util/TchapRoomUtils");

describe("RoomGeneralContextMenu", () => {
    const ROOM_ID = "!123:matrix.org";

    let room: Room;
    let mockClient: MatrixClient;

    let onFinished: () => void;

    const mockedTchapRoomUtils = mocked(TchapRoomUtils);

    function getComponent(props?: Partial<RoomGeneralContextMenuProps>) {
        return render(
            <MatrixClientContext.Provider value={mockClient}>
                <RoomGeneralContextMenu
                    room={room}
                    onFinished={onFinished}
                    {...props}
                    managed={true}
                    mountAsChild={true}
                    left={1}
                    top={1}
                    chevronFace={ChevronFace.Left}
                />
            </MatrixClientContext.Provider>,
        );
    }

    beforeEach(() => {
        jest.clearAllMocks();

        stubClient();
        mockClient = mocked(MatrixClientPeg.safeGet());

        room = new Room(ROOM_ID, mockClient, mockClient.getUserId() ?? "", {
            pendingEventOrdering: PendingEventOrdering.Detached,
        });

        const dmRoomMap = {
            getUserIdForRoomId: jest.fn().mockReturnValue(null),
        } as unknown as DMRoomMap;
        DMRoomMap.setShared(dmRoomMap);

        jest.spyOn(RoomListStore.instance, "getTagsForRoom").mockReturnValueOnce([
            DefaultTagID.DM,
            DefaultTagID.Favourite,
        ]);

        room.updateMyMembership(KnownMembership.Join);
        jest.spyOn(room, "canInvite").mockReturnValue(true);

        onFinished = jest.fn();
    });

    afterEach(async () => {
        await clearAllModals();
    });

    it("renders the default context menu", async () => {
        const { container } = getComponent({});
        expect(container).toMatchSnapshot();
    });

    it("able the copy link button when joinrule is public", () => {
        mockedTchapRoomUtils.getRoomJoinRule.mockImplementation(() => JoinRule.Public);

        const { container } = getComponent({});

        logRoles(container);

        const roomLinkButton = screen.getByRole("menuitem", { name: "Copy room link" });

        expect(roomLinkButton).not.toHaveAttribute("aria-disabled");
    });

    it("disable the copy link button when joinrule is not public", () => {
        mockedTchapRoomUtils.getRoomJoinRule.mockImplementation(() => JoinRule.Invite);

        getComponent({});

        const roomLinkButton = screen.getByRole("menuitem", { name: "Copy room link" });

        expect(roomLinkButton).toHaveAttribute("aria-disabled", "true");
    });
});
