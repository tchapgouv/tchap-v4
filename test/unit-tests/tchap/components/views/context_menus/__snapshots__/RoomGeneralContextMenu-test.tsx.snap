// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RoomGeneralContextMenu renders the default context menu 1`] = `
<div>
  <div
    class="mx_ContextualMenu_wrapper"
    style="top: 1px; left: 1px;"
  >
    <div
      class="mx_ContextualMenu_background"
    />
    <div
      class="mx_ContextualMenu mx_ContextualMenu_withChevron_left"
      role="menu"
    >
      <div
        class="mx_ContextualMenu_chevron_left"
      />
      <ul
        class="mx_IconizedContextMenu mx_RoomGeneralContextMenu mx_IconizedContextMenu_compact"
        role="none"
      >
        <div
          class="mx_IconizedContextMenu_optionList mx_IconizedContextMenu_optionList_notFirst"
        >
          <li
            aria-label="Mark as unread"
            class="mx_AccessibleButton mx_IconizedContextMenu_item"
            role="menuitem"
            tabindex="0"
          >
            <span
              class="mx_IconizedContextMenu_icon mx_RoomGeneralContextMenu_iconMarkAsUnread"
            />
            <span
              class="mx_IconizedContextMenu_label"
            >
              Mark as unread
            </span>
          </li>
          <div
            aria-checked="false"
            aria-label="Favourite"
            class="mx_AccessibleButton mx_IconizedContextMenu_item"
            role="menuitemcheckbox"
            tabindex="-1"
          >
            <span
              class="mx_IconizedContextMenu_icon mx_RoomGeneralContextMenu_iconStar"
            />
            <span
              class="mx_IconizedContextMenu_label"
            >
              Favourite
            </span>
            <span
              class="mx_IconizedContextMenu_icon mx_IconizedContextMenu_unchecked"
            />
          </div>
          <div
            aria-checked="false"
            aria-label="Low Priority"
            class="mx_AccessibleButton mx_IconizedContextMenu_item"
            role="menuitemcheckbox"
            tabindex="-1"
          >
            <span
              class="mx_IconizedContextMenu_icon mx_RoomGeneralContextMenu_iconArrowDown"
            />
            <span
              class="mx_IconizedContextMenu_label"
            >
              Low Priority
            </span>
            <span
              class="mx_IconizedContextMenu_icon mx_IconizedContextMenu_unchecked"
            />
          </div>
          <li
            aria-label="Invite"
            class="mx_AccessibleButton mx_IconizedContextMenu_item"
            role="menuitem"
            tabindex="-1"
          >
            <span
              class="mx_IconizedContextMenu_icon mx_RoomGeneralContextMenu_iconInvite"
            />
            <span
              class="mx_IconizedContextMenu_label"
            >
              Invite
            </span>
          </li>
          <li
            aria-label="Copy room link"
            class="mx_AccessibleButton mx_IconizedContextMenu_item"
            role="menuitem"
            tabindex="-1"
          >
            <span
              class="mx_IconizedContextMenu_icon mx_RoomGeneralContextMenu_iconCopyLink"
            />
            <span
              class="mx_IconizedContextMenu_label"
            >
              Copy room link
            </span>
          </li>
          <li
            aria-label="Settings"
            class="mx_AccessibleButton mx_IconizedContextMenu_item"
            role="menuitem"
            tabindex="-1"
          >
            <span
              class="mx_IconizedContextMenu_icon mx_RoomGeneralContextMenu_iconSettings"
            />
            <span
              class="mx_IconizedContextMenu_label"
            >
              Settings
            </span>
          </li>
        </div>
        <div
          class="mx_IconizedContextMenu_optionList mx_IconizedContextMenu_optionList_notFirst mx_IconizedContextMenu_optionList_red"
        >
          <li
            aria-label="Leave"
            class="mx_AccessibleButton mx_IconizedContextMenu_option_red mx_IconizedContextMenu_item"
            role="menuitem"
            tabindex="-1"
          >
            <span
              class="mx_IconizedContextMenu_icon mx_RoomGeneralContextMenu_iconSignOut"
            />
            <span
              class="mx_IconizedContextMenu_label"
            >
              Leave
            </span>
          </li>
        </div>
      </ul>
    </div>
  </div>
</div>
`;
